sklearn_numba_dpex/tests/test_kmeans.py

```diff
@@ -176,3 +177,113 @@ def test_inertia(dtype):
+def test_select_samples_far_from_centroid_kernel(dtype):
+    _fail_if_no_dtype_support(pytest.xfail, dtype)
+
+    # The following array, when sorted, reads:
+    # [-1.0, 9.0, 9.0, 20.22, 20.22, 20.22, 20.22, 23.0, 23.0, 30.0]
+    # So we expect 3 values above threshold: 23.0, 23.0, 30.0
+    # and 2 values equal to threshold.
+    distance_to_centroid = dpctl.tensor.from_numpy(
+        np.array(
+            [
+                20.22,  # == threshold
+                20.22,
+                -1.0,
+                23.0,  # idx = 3
+                20.22,
+                9.0,
+                20.22,
+                30.0,  # idx = 7
+                9.0,
+                23.0,  # idx = 9
+            ],
+            dtype=dtype,
+        )
+    )
+
+    # NB: we wrap everything in dpctl tensors and select the threshold with __getitem__
+    # on a dpctl.tensor because loading of numpy arrays to and from dpctl.tensor arrays
+    # seems to create numerical errors that break the kernel
+    # (the condition ==threshold stops being evaluated properly) when using float64
+    # precision ?
+    # TODO: write a minimal reproducer and open an issue if confirmed
+    threshold = distance_to_centroid[0:1]  # == 20.22
+
+    n_selected = 5
+    n_samples = len(distance_to_centroid)
+    work_group_size = 4
+    select_samples_far_from_centroid_kernel = (
+        make_select_samples_far_from_centroid_kernel(
+            n_selected, n_samples, work_group_size
+        )
+    )
+
+    # NB: the values used to initialize the output array does not matter, 100 is chosen
+    # here for readability, but `dpctl.empty` is also possible.
+    selected_samples_idx = (dpctl.tensor.ones(sh=10, dtype=np.int32) * 100).get_array()
+
+    n_selected_gt_threshold = dpctl.tensor.zeros(sh=1, dtype=np.int32)
+    n_selected_eq_threshold = dpctl.tensor.ones(sh=1, dtype=np.int32)
+    select_samples_far_from_centroid_kernel(
+        distance_to_centroid,
+        threshold,
+        selected_samples_idx,
+        n_selected_gt_threshold,
+        n_selected_eq_threshold,
+    )
+
+    # NB: the variable n_selected_eq_threshold is always one unit above the true value
+    # It is only used as an intermediary variable in the kernel and is not used
+    # otherwise
+    n_selected_eq_threshold = int(n_selected_eq_threshold[0] - 1)
+    n_selected_gt_threshold = int(n_selected_gt_threshold[0])
+    selected_samples_idx = dpctl.tensor.asnumpy(selected_samples_idx)
+
+    # NB: the exact number of selected values equal to the threshold and the
+    # corresponding selected indexes in the input array can change depending on
+    # concurrency. We only check conditions for success, those does not depend on the
+    # variability.
+    assert n_selected_gt_threshold == 3
+    assert n_selected_eq_threshold >= 2
+    assert set(selected_samples_idx[:3]) == {3, 7, 9}
+    assert set(selected_samples_idx[-n_selected_gt_threshold:]).issubset({0, 1, 4, 6})
```

Comment:
If I am not mistaken, the second assertion should use `n_selected_gt_threshold` instead of `-n_selected_eq_threshold`.

```suggestion
    expected_gt_threshold_indices = {3, 7, 9}
    actual_gt_threshold_indices = set(
        selected_samples_idx[:n_selected_gt_threshold]
    )  # stored at the beginning
    assert actual_gt_threshold_indices == expected_gt_threshold_indices
 
    expected_eq_threshold_indices = {0, 1, 4, 6}
    actual_eq_threshold_indices = set(
        selected_samples_idx[-n_selected_eq_threshold:]
    )  # stored at the end
    assert actual_eq_threshold_indices.issubset(expected_eq_threshold_indices)
```

not sure about by manual code formatting